File: DDD/SPAAngularJSMVC/Solucao/MVCAngularJS.Web.Mvc/Controllers/CategoriaController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using MVCAngularJS.Aplicacao.Dto;
using MVCAngularJS.Aplicacao.Fachada;
using MVCAngularJS.Web.Mvc.Models;
using Newtonsoft.Json;

namespace MVCAngularJS.Web.Mvc.Controllers
{
	public class CategoriaController : Controller
	{
        ICategoriaApp app;
		//
		// GET: /Categoria/
		public ActionResult Index()
		{
			return View();
		}

        public CategoriaController(ICategoriaApp app)
        {
            this.app = app;
        }

		[HttpPost]
		public JsonResult CadastrarCategoria(CategoriaCadastroModel model)
		{
            if (model == null || String.IsNullOrEmpty(model.Nome))
            {
                return Json(new { successo = false, textoDeResposta = "Objet não pode ser vazio" }, JsonRequestBehavior.AllowGet);
            }

            try
            {                
                CategoriaDto dto = new CategoriaDto()
                {
                    Nome = model.Nome
                };
                
                app.Cadastrar(dto);

                return Json(new { successo = true, textoDeResposta = String.Format("Categoria {0} cadstrada com sucesso!", dto.Nome) }, JsonRequestBehavior.AllowGet);

                

            }catch(Exception ex){
                return Json(new { successo = false, textoDeResposta = String.Format("Erro ao cadastrar categoria {0}", ex.Message) }, JsonRequestBehavior.AllowGet);
            }
            
		}

		[HttpGet]        
		public JsonResult ListarTodasAsCategorias()
		{            
            ICollection<CategoriaDto> dtos;

            try
            {
                dtos = app.BuscarTodos();

                if (dtos != null && dtos.Count > 0)
                {
                    return Json(dtos, JsonRequestBehavior.AllowGet);
                }
                else
                {
                    throw new HttpException(403,"Nenhuma Categoria encontrada!");
                }
            }
            catch (Exception ex)
            {

                return Json(new { success = false, responseText = String.Format("Erro ao Listar todas as categorias {0}", ex.Message) }, JsonRequestBehavior.AllowGet);
            }
                        
		}

        [HttpGet]
        public JsonResult BuscarCategoriaPorId(int id)
        {
            try
            {
                CategoriaDto dto = app.BuscarPorId(id);

                if(dto != null){
                    return Json(dto, JsonRequestBehavior.AllowGet);
                }
                else
                {
                    throw new HttpException(403, "Nenhuma Categoria encontrada!");
                }

            }catch(Exception ex){
                return Json(new { success = false, responseText = String.Format("Erro ao Listar a categoria pelo id {0}: {1}", ex.Message, id) }, JsonRequestBehavior.AllowGet);
            }
        }
	}
}

File: DDD/SPAAngularJSMVC/Solucao/MVCAngularJS.Web.Mvc/Angular/Controllers/categoria_controller.js
aplicacao.controller(
    'CategoriaController',
    function($scope,$http){
        
        //Inicializa Model
        $scope.CategoriaConsultaModel = {
            Id: '',
            Nome :''
        };

        $scope.CategoriaCadastroModel = {
            Id: '',
            Nome :''
        };

        $scope.CadastrarCategoria = function () {
            $scope.mensagem = "Aguarde... processando informação";

            $http.post("http://localhost:12108/Categoria/CadastrarCategoria", $scope.CategoriaCadastroModel)
                .success
                (
                    function (resposta) {
                        $scope.mensagem = msg.textoResposta;

                        //Limpar os campos
                        $scope.CategoriaCadastroModel.Nome = '';
                    }
                )
                .error(
                        function (e) {
                            $scope.mensagem = e; //imprimir..
                        }
                    );                
        };

        $scope.ConsultarPorId = function (id)
        {
            $http.get("http://localhost:12108/Categoria/BuscarCategoriaPorId/" + id)
                .success
                (
                    function (categoria) {
                        $scope.CategoriaConsultaModel.Id = categoria.Id;
                        $scope.CategoriaConsultaModel.Nome = categoria.Nome;
                    }
                )
                .error
                (
                    function (e) {
                        $scope.mensagem = e;
                    }
                );
        }

        $scope.ListarTodasAsCategorias = function()
        {
            $http.get("http://localhost:12108/Categoria/ListarTodasAsCategorias")
                .success
                (
                    function (resultado)
                    {
                        $scope.categorias = resultado;
                    }
                )
                .error(
                        function (e) {
                            $scope.mensagem = e;
                        }
                    );
        }

    });